### DISCUSSION TITLE
`WARNING  Unable to load module ... for options validation`

### DISCUSSION THREAD

**Discussion by Vanav** [ansible/ansible-lint#2861] [Q&A]
`ansible-lint` displays warnings:
```
WARNING: PATH altered to include /opt/pipx/venvs/ansible-lint/bin :: This is usually a sign of broken local setup, which can cause unexpected behaviors.
...
WARNING  Unable to load module snap at roles/base/tasks/main.yml:32 for options validation
WARNING  Unable to load module timezone at roles/base/tasks/main.yml:60 for options validation
WARNING  Unable to load module zabbix_group at roles/zabbix/tasks/main.yml:165 for options validation
WARNING  Unable to load module zabbix_host at roles/zabbix/tasks/main.yml:172 for options validation
WARNING  Unable to load module zabbix_hostmacro at roles/zabbix/tasks/main.yml:209 for options validation
WARNING  Unable to load module synchronize at roles/deploy/tasks/update-code/rsync_direct.yml:13 for options validation
WARNING  Unable to load module community.general.snap at roles/nginx/tasks/main.yml:60 for options validation
WARNING  Unable to load module postgresql_user at roles/postgres-env/tasks/main.yml:5 for options validation
WARNING  Unable to load module postgresql_db at roles/postgres-env/tasks/main.yml:12 for options validation
WARNING  Unable to load module postgresql_ext at roles/postgres-env/tasks/main.yml:19 for options validation
WARNING  Unable to load module git_config at roles/base/tasks/Debian.yml:291 for options validation
WARNING  Unable to load module locale_gen at roles/base/tasks/Debian.yml:370 for options validation
WARNING  Unable to load module authorized_key at roles/unix-users/tasks/main.yml:110 for options validation
```
I have installed ansible and ansible-lint in pipx environments, as recommended:
```
# ansible-lint --version
WARNING: PATH altered to include /opt/pipx/venvs/ansible-lint/bin :: This is usually a sign of broken local setup, which can cause unexpected behaviors.
ansible-lint 6.10.0 using ansible 2.14.0

# pipx list
venvs are in /opt/pipx/venvs
apps are exposed on your $PATH at /opt/pipx/bin
   package ansible 7.1.0, installed using Python 3.10.6
    - ansible
    - ansible-community
    - ansible-config
    - ansible-connection
    - ansible-console
    - ansible-doc
    - ansible-galaxy
    - ansible-inventory
    - ansible-playbook
    - ansible-pull
    - ansible-test
    - ansible-vault
   package ansible-lint 6.10.0, installed using Python 3.10.6
    - ansible-lint
```
Playbook has no issues. I've tried to use FQCN for modules, but it doesn't change anything: same warnings, but with FQCN paths.

**Comment by netsandbox** [comment]
I'm also see this warnings, but other than @Vanav report, I actually activated the venv.

```shell
$ ansible-lint
WARNING: ansible-lint is no longer tested under Python 3.8 and will soon require 3.9. Do not report bugs for this version.
WARNING  Unable to load module ansible.windows.win_copy at playbooks/ssl_cert_windows.yaml:25 for options validation
WARNING  Unable to load module ansible.windows.win_certificate_store at playbooks/ssl_cert_windows.yaml:31 for options validation
WARNING  Unable to load module ansible.windows.win_file at playbooks/ssl_cert_windows.yaml:37 for options validation
WARNING  Unable to load module community.windows.win_disk_facts at playbooks/facts.yaml:23 for options validation
```

```shell
$ ansible-galaxy collection list

# /home/ansible/collections/ansible_collections
Collection            Version
--------------------- -------
ansible.posix         1.4.0  
ansible.windows       1.12.0 
community.general     6.0.1  
community.hashi_vault 4.0.0  
community.vmware      2.10.2 
community.windows     1.11.1 
```

**Comment by nickkeyzer** [comment]
Possibly related but I'm seeing the same symptoms as described here for all Windows modules: https://github.com/ansible/ansible-lint/issues/2831

**Comment by hemberger** [comment]
I don't see this issue when I run ansible-lint locally (v6.10.0), but I do see it in my Actions workflow. I've reported it here: https://github.com/ansible/ansible-lint-action/issues/99#issuecomment-1365487875.

**Reply by ssbarnea** [reply]
The action is using a container, so it will not have access to your locally installed collections. That might explain why action users get these errors. Still; if they have a requirements.yml, the action should install the collections inside the container (given that they did not enable offline mode).

**Comment by Schachte** [comment]
This is affecting me as well. 

### requirements.yaml

```
---
collections:
  - name: community.docker
    source: https://galaxy.ansible.com
```

In my action, I pre-install this.

```
name: Ansible Deployment

on:
  workflow_dispatch:

jobs:
  linting:
    name: List Ansible Files
    runs-on: ubuntu-22.04

    steps:
    - name: Checkout Sources
      uses: actions/checkout@v2

    - name: Install Ansible
      run: sudo apt-get install -y ansible

    - name: Install Ansible-Galaxy Requirements
      run: ansible-galaxy install -r requirements.yaml

    - name: List Installed Ansible-Galaxy Requirements
      run: ansible-galaxy collection list

    - name: Linting
      uses: ansible/ansible-lint-action@dd225474d15635904348151b9e694472ce0f62a8
      with:
        path: ansible/
```

I see the output of the `collection list` showing exactly what I installed, but the Github Action fails to resolve with the following:


```
WARNING  Unable to load module community.docker.docker_compose at ansible/playbooks/roles/nginx/tasks/nginx.yaml:2 for options validation
[10](https://github.com/Schachte/homelab/actions/runs/3813289308/jobs/6486951382#step:7:11)
WARNING  Unable to load module community.docker.docker_compose at ansible/playbooks/roles/nginx/tasks/nginx.yaml:7 for options validation
[11](https://github.com/Schachte/homelab/actions/runs/3813289308/jobs/6486951382#step:7:12)
WARNING  Unable to load module community.docker.docker_compose at ansible/playbooks/roles/nginx/tasks/nginx.yaml:12 for options validation
[12](https://github.com/Schachte/homelab/actions/runs/3813289308/jobs/6486951382#step:7:13)
WARNING  Unable to resolve FQCN for module community.docker.docker_compose
```

**Comment by ssbarnea** [comment]
I will be converting this to a discussion as there is no specific bug identified. I will fine a bug once we identify a clear bug. When it comes to use of GHA action, there are several things people need to remember because that is running inside a containers, so doing any preparatory work is unlikely to help in a any way.

**Comment by mbwhite** [comment]
I'm seeing this consistently github-actions or local.  I find Pythons environment setup is at the best of time confusing; 

It looks like when we're lining the roles that refer to the modules in the collection - it's the FQCN that can't be resolved.

**Reply by ssbarnea** [reply]
Pytho venvs are confusing as people forget to activate the venv and try to run executables from within the venv. They appear to work... until these executables try to execute another sibling executable which is not in PATH. They even fail with command not found or worse, they call another different version of that command that happens to be installed at system level.

Shortly, venvs are very good and reliable, for those that do understand how they work ;)

**Comment by interdependence** [comment]
This starts happening for me when upgrading to v6.12.0 from v6.11.0 and running ansible-lint from pre-commit. Interestingly it doesn't happen when I run ansible-lint >v6.12.0 manually.

**Reply by ssbarnea** [reply]
I wonder if precommit run hides or changes HOME directory as that could prevent ansible find collections from user directory.

**Reply by bluikko** [reply]
It used to work for me too, after 6.12.2 everything outside of `ansible.builtin` is broken. Are you using offline mode? That has been a problem in the past and I am using offline mode.

**Comment by skokhanovskiy** [comment]
Managed to write minimal Dockerfile with PoC of this issue. There are no virtual environments in it, but loading modules doesn't work and produces warnings

```
WARNING  Unable to load module community.general.pamd at playbook.yml:7 for options validation
WARNING  Unable to resolve FQCN for module community.general.pamd
```

https://github.com/ansible/ansible-lint/issues/2969#issuecomment-1419000179

**Comment by hollow** [comment]
I am following up on this discussion from #2969.

The [error reported in the above issue](https://github.com/ansible/ansible-lint/issues/2969#issuecomment-1415728643) seems to have started once I removed my globally installed ansible/ansible-lint and switched to a poetry-based installation of ansible/ansible-lint. While moving to poetry I have also installed ansible-core only (i.e. no ansible package with collections) because these are exclusively managed and installed from requirements.yaml in our case.

**Comment by nre-ableton** [comment]
I originally created https://github.com/ansible/ansible-lint/issues/2831#event-8134410171, which was closed as `wontfix`. If I had to take a guess, this issue is probably related to the handling of `ANSIBLE_COLLECTIONS_PATH`, which has always been problematic in ansible-lint.

FWIW, we are using `--offline`, but omitting it results in the same behavior.

**Reply by bluikko** [reply]
Yeah offline has been broken in many versions - and I suspect that's the issue again. Going back to 6.11 to get it working.

**Comment by bluikko** [comment]
No idea how this is "not a bug" or was moved to a discussion. Everything used to work in my system prior to update from 6.11.x to 6.12.2 and now everything outside of `ansible.builtin` is "Unable to load" with the exact same symptoms. It is the update that broke ansible-lint - nothing else has been changed...

Edited to add: I have in the past have problems with the offline mode, and the numerous releases trying to get it fixed. I am still continuing to use it and I wonder if this problem is because of offline mode.

**Reply by nre-ableton** [reply]
I believe that this was classified as "not a bug" because the maintainers are [unable to reproduce it](https://github.com/ansible/ansible-lint/issues/2831#issuecomment-1368489519) themselves.

I also suspect offline mode is responsible here, but as stated in my comment above, I get the warming regardless.

**Reply by bluikko** [reply]
Well I went back to 6.11 and everything works. Pretty sure it's offline mode, it has been broken in every new minor release going back x releases.

**Comment by qwestduck** [comment]
At least some of the behavior in #2828 and #2831 can be explained by most of the modules in ansible.windows and community.windows being implemented in powershell rather than python. See https://github.com/ansible-collections/ansible.windows/blob/main/plugins/modules/win_ping.ps1 as an example. Attempting to load a python spec from a powershell source (https://github.com/ansible/ansible-lint/blob/v6.14.0/src/ansiblelint/rules/args.py#L111) fails and logs the "Unable to load..." warning with no further option validation (https://github.com/ansible/ansible-lint/blob/v6.14.0/src/ansiblelint/rules/args.py#L121).

If it is a bug that these modules are not validated, then this test asserts the correct behavior: https://github.com/ansible/ansible-lint/compare/main...qwestduck:ansible-lint:test/2831-rules-args-windows-options-validation. I would think the way to get there might be parsing option constraints out of the python sidecar docfiles that accompany these powershell modules (for example https://github.com/ansible-collections/ansible.windows/blob/main/plugins/modules/win_ping.py), but that would add a lot to the rule for only a small set of modules.

**Comment by regner** [comment]
So has anyone have a solution for this? Struggling to understand how this isn't a bug, or if it isn't what I am doing wrong.

To make sure I am working with the correct collections, and just generally that paths are setup correctly in Ansible I set my ansible.cfg to only load collections from my ansible home:
```cfg
[defaults]
collections_scan_sys_path = False
collections_path = ~/.ansible/collections
```

I verified this was working, and not loading anything, by attempting to list collections:
```bash
$ ansible-galaxy collection list
usage: ansible-galaxy [-h] [--version] [-v] TYPE ...

Perform various Role and Collection related operations.

positional arguments:
  TYPE
    collection   Manage an Ansible Galaxy collection.
    role         Manage an Ansible Galaxy role.

options:
  --version      show program's version number, config file location, configured module search path, module location, executable location and exit
  -h, --help     show this help message and exit
  -v, --verbose  verbose mode (-vvv for more, -vvvv to enable connection debugging)
ERROR! - None of the provided paths were usable. Please specify a valid path with --collections-path
```

I then tried to run ansible lint. It failed being unable to load a large number of modules:
```bash
$ ansible-lint
WARNING  Listing 57 violation(s) that are fatal
syntax-check[specific]: couldn't resolve module/action 'community.docker.docker_prune'. This often indicates a misspelling, missing collection, or incorrect module path.
playbooks/docker-prune.yaml:6:7
```

I then installed all my collections and roles, and listed them out again:
```bash
$ ansible-galaxy collection list

# /home/vscode/.ansible/collections/ansible_collections
Collection            Version
--------------------- -------
ansible.posix         1.5.1  
ansible.windows       1.13.0 
chocolatey.chocolatey 1.4.0  
community.docker      3.4.2  
community.general     6.4.0  
community.windows     1.12.0 
```

I then ran ansible lint again and got the warnings others have been talking about, all of them Windows related:
```bash
$ ansible-lint
WARNING  Unable to load module community.windows.win_partition at roles/windows_base/tasks/disks.yaml:2 for options validation
WARNING  Unable to load module ansible.windows.win_service at roles/windows_base/tasks/service.yaml:2 for options validation
WARNING  Unable to load module ansible.windows.win_domain_membership at roles/windows_base/tasks/domain.yaml:2 for options validation
```

What's interesting is that ansible.windows.win_template does not cause a warning, and it is used within our roles. That module does not have a .ps1 file in the collection, it only has a .py file.

**Comment by gone-for-coding** [comment]
I think I was able to narrow the issue down to the new args rule and opened a new bug report with a minimal example:

https://github.com/ansible/ansible-lint/issues/3200
